go-controller/pkg/node/gateway_init.go

```diff
@@ -534,3 +535,38 @@ func CleanupClusterNode(name string) error {
 
 	return nil
 }
+
+func (nc *DefaultNodeNetworkController) updateGatewayMAC(link netlink.Link) error {
+	if nc.gateway.GetGatewayBridgeIface() != link.Attrs().Name {
+		return nil
+	}
+
+	node, err := nc.watchFactory.GetNode(nc.name)
+	if err != nil {
+		return err
+	}
+	l3gwConf, err := util.ParseNodeL3GatewayAnnotation(node)
+	if err != nil {
+		return err
+	}
+
+	if l3gwConf == nil || l3gwConf.MACAddress.String() == link.Attrs().HardwareAddr.String() {
+		return nil
+	}
+	// MAC must have changed, update node
+	nc.gateway.SetDefaultGatewayBridgeMAC(link.Attrs().HardwareAddr)
+	if err := nc.gateway.Reconcile(); err != nil {
+		return fmt.Errorf("failed to reconcile gateway for MAC address update: %w", err)
+	}
+	nodeAnnotator := kube.NewNodeAnnotator(nc.Kube, node.Name)
+	l3gwConf.MACAddress = link.Attrs().HardwareAddr
+	if err := util.SetL3GatewayConfig(nodeAnnotator, l3gwConf); err != nil {
+		return fmt.Errorf("failed to update L3 gateway config annotation for node: %s, error: %w", node.Name, err)
+	}
+	if err := nodeAnnotator.Run(); err != nil {
+		return fmt.Errorf("failed to set node %s annotations: %w", nc.name, err)
+	}
```

Comment:
How do the node port watcher flows get updated? is that through the annotations?

Reply:
nice catch! I missed this. I'll try to figure out the easiest way to trigger this update.